File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandBuilders/BroadcastCommandBuilder.java
package bgu.spl171.net.impl.TFTP.CommandBuilders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.ReadRequest;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.StringEncoderDecoder;

public class BroadcastCommandBuilder implements CommandBuilder {
    Boolean isAdded = null;
    StringEncoderDecoder stringEncoderDecoder = new StringEncoderDecoder();

    @Override
    public Command decodeNextByte(byte nextByte) {
        if (isAdded != null) {
            String fileName = stringEncoderDecoder.decodeNextByte(nextByte);
            if (fileName != null) {
                return new ReadRequest(fileName);
            }
        }
        else {
            isAdded = nextByte == 1;
        }
        return null;
    }
}

File: java to c++ builders/BroadcastCommandBuilder.cpp
#include "BroadcastCommandBuilder.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandBuilders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::ReadRequest;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::StringEncoderDecoder;

						Command *BroadcastCommandBuilder::decodeNextByte(char nextByte)
						{
							if (isAdded)
							{
								std::wstring fileName = stringEncoderDecoder->decodeNextByte(nextByte);
								if (fileName != L"")
								{
									return new ReadRequest(fileName);
								}
							}
							else
							{
								isAdded = nextByte == 1;
							}
							return nullptr;
						}
					}
				}
			}
		}
	}
}
